### DISCUSSION TITLE
Select Component : Dynamic options is not working with searchable

### DISCUSSION THREAD

**Discussion by maystro** [filamentphp/filament#11612] [General]
Iam trying to load data dynamically inside Select Dropdown, When i try to use ->searchable(), it does not load any thing
note that, without searchable() it works well,
here is my code

                Select::make('requests')
                    ->label('Request type')
                    ->live()
                    ->searchable()
                    ->disabled(fn(Forms\Get $get): bool => !(self::getEmployee()))
                    ->options(function (Forms\Get $get, $state) {
                        if (self::getEmployee()) {
                            $allowance_rule = self::$employee->allowance_rule;
                            $allowance_limits = $allowance_rule->limits;
                            return $allowance_limits->pluck('allowance.name', 'id');
                        } else
                        {
                            return [];
                        }
                    })
                    ->columnSpan(4),

**Comment by zeeshantariq08** [comment]
you can add preload()

**Reply by maystro** [reply]
i did, but still not work, I spent 6 hours try to find a way

**Reply by leandrocfe** [reply]
`->searchable(fn() => self::getEmployee())`
?

**Reply by maystro** [reply]
sadly, I did that now, not work
and Iam sure self::getEmployee() is not null, I dump it and it has data

**Reply by leandrocfe** [reply]
Are you using the panel builder or the form builder only?

**Reply by maystro** [reply]
Panel builder, not custom page or widget

**Reply by maystro** [reply]
I found the same issue in Stackoverflow with no answer
https://stackoverflow.com/questions/77239006/dynamic-select-dropdown-in-filament-not-working

**Reply by leandrocfe** [reply]
https://discord.com/channels/883083792112300104/1199415958045982870/1199718177408557096

**Reply by maystro** [reply]
link to empty page!

**Comment by maystro** [comment]
this is the whole class


namespace App\Filament\Hr\Resources;


use App\Filament\Helper\EmpClass;
use App\Filament\Helper\FilamentHelper;
use App\Filament\Hr\Resources\EmployeeAllowanceResource\Pages;
use App\Filament\Hr\Resources\EmployeeAllowanceResource\RelationManagers;
use App\Filament\Hr\Resources\EmployeeAllowanceResource\Widgets\FindEmployeeWidget;
use App\Http\Classes\HrSidebarGroups;
use App\Models\Allowance;
use App\Models\AllowanceLimit;
use App\Models\AllowanceRule;
use App\Models\Employee;
use App\Models\EmployeeAllowance;
use Carbon\PHPStan\AbstractMacro;
use Faker\Provider\ar_EG\Text;
use Filament\Actions\Action;
use Filament\Forms;
use Filament\Forms\Components\Actions;
use Filament\Forms\Components\Checkbox;
use Filament\Forms\Components\Section;
use Filament\Forms\Components\Select;
use Filament\Forms\Components\TextInput;
use Filament\Forms\Form;
use Filament\Forms\Set;
use Filament\Infolists\Components\TextEntry;
use Filament\Resources\Resource;
use Filament\Tables;
use Filament\Tables\Table;
use http\Encoding\Stream\Deflate;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Collection;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletingScope;
use Illuminate\Session\SessionManager;
use JetBrains\PhpStorm\NoReturn;
use PHPUnit\Event\Code\Throwable;
use function PHPUnit\Framework\throwException;

class EmployeeAllowanceResource extends Resource
{
    protected static ?string $model = EmployeeAllowance::class;
    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';
    protected static ?string $navigationGroup = HrSidebarGroups::HRForms;
    protected static ?int $navigationSort = 1;
    protected static ?string $pluralLabel = "أجازات وأذونات";
    protected static ?string $modelLabel = "طلب اذن";
    protected static ?string $navigationLabel = 'طلبات الأجازات والأذونات';
    public static ?Employee $employee = null;
    public static ?AllowanceLimit $request = null;
    public static string|null $requestId = null;
    static function updateUI(Set $set, Forms\Get $get): void
    {
        if (self::getEmployee()) {
            $record = self::getEmployee();
            if ($record) {
                $allowance_rule = $record->allowance_rule;
                $allowance_limits = $allowance_rule->limits;
                $allowance_limit = $allowance_limits->where('id','=',self::$requestId)->first();
                //$set('code', $record->code);
                //$set('allowance_id', self::$requestId);
                if (static::$requestId)
                {
                    session()->put('request',self::$request);
                    $set('allowance_unit', AllowanceLimit::translateUnit($allowance_limit->allowance_unit));
                }
            }
        }
    }
    public static function getEmployee(): ?Employee
    {
        if(session('employee')!=null)
        {
            static::$employee = session()->get('employee');
            return static::$employee;
        }
        else return null;
    }
    public static function getRequest(): ?AllowanceLimit
    {
        if(session()->get('request')!=null)
        {
            static::$request = session()->get('request');
            return static::$request;
        }
        else return null;
    }

    /**
     * @param mixed $empId
     * @param Set $set
     * @return Employee|Model|null
     */
    private static function fillEmployee(mixed $empId, Set $set): null|Employee|\Illuminate\Database\Eloquent\Model
    {
        self::$employee = Employee::query()->where('code', '=', $empId)->first();

        if (self::$employee) {
            $set('employee_name', self::$employee->employee_name);
            $set('employee_code', self::$employee->code);
            $set('code', self::$employee->code); //hidden

            session()->put('employee', self::$employee);

            return self::$employee;
        } else return null;
    }
    public static function refreshEmployee(Set $set, Forms\Get $get, $component, $state): \Illuminate\Database\Eloquent\Model|Employee|Builder|\Exception|null
    {
        if (filled($get('employee_code'))) {
            $empId = $get('employee_code');
            return self::fillEmployee($empId, $set);
        }
        elseif (filled($get('employee')))
        {
            $empId = $get('employee');
            return self::fillEmployee($empId, $set);
        }
        else {
            return new \Exception('Employee CODE not found');
        }
    }
    static function createEmployeePhotoSchema(): Forms\Components\Section
    {
        return Forms\Components\Section::make('صورة شخصية')->schema([
            Forms\Components\FileUpload::make('image')
                ->alignJustify()
                ->alignCenter()
                ->hiddenLabel()
                ->imageEditor()
                ->imageCropAspectRatio('1:1')
                ->maxSize(1024) //Kb 1024k = 1M
                ->directory('uploads/employees/images')
                ->validationMessages([
                    'max' => 'Max file is 1M'
                ])
        ]);
    }
    static function createSearchSection(): Section
    {
        return Section::make()->schema([
            Forms\Components\Radio::make('searchMethod')
                ->options(['search_ByCode' => 'بحث بالكود', 'search_ByName' => 'بحث بالإسم'])
                ->live()
                ->default('search_ByCode')
                ->hiddenLabel()
                ->inline(),

            //** SEARCH BY CODE */
            Forms\Components\Group::make()->schema([
                TextInput::make('employee_code')
                    ->live()
                    ->afterStateUpdated(function (Set $set, Forms\Get $get, $component, $state) {
                        static::refreshEmployee($set, $get, $component, $state);
                    })
                    ->label('كود الموظف')
                    ->suffixAction(Actions\Action::make('find')
                        ->icon('heroicon-o-magnifying-glass')
                        ->action(function (Set $set, Forms\Get $get, $component, $state) {
                            static::refreshEmployee($set, $get, $component, $state);
                        })
                    ),

                TextInput::make('employee_name')
                    ->label('اسم الموظف')
                    ->readOnly(true)
            ])->hidden(fn(Forms\Get $get): bool => $get('searchMethod') == 'search_ByName'),

            //** SEARCH BY NAME */
            Forms\Components\Group::make()->schema([
                Select::make('employee')
                    ->label('اسم الموظف')
                    ->options(fn() => Employee::all()->pluck('employee_name', 'code'))
                    ->searchable()
                    ->optionsLimit(5)
                    ->live()
                    ->afterStateUpdated(function (Set $set, Forms\Get $get, $component, $state) {
                        static::refreshEmployee($set, $get, $component, $state);
                    })

            ])->hidden(fn(Forms\Get $get): bool => $get('searchMethod') == 'search_ByCode'),

            Actions::make([
                Actions\Action::make('star')
                    ->icon('heroicon-m-star')
                    ->requiresConfirmation()
                    ->action(function () {
                    }),
            ])->fullWidth(),

        ])->columnSpan(1);
    }

    private function createAllowanceLimit(): Section
    {
        return Section::make()->schema([
            TextInput::make('limit')
                ->disabled()
                ->label('الرصيد المتاح'),

            TextInput::make('LastAllowance')
                ->disabled()
                ->label('تاريخ اخر اذن'),
        ]);
    }

    static function createRequestForm(): Section
    {
        return Section::make()->schema([

            Forms\Components\Group::make()->schema([
                Select::make('requests')
                    ->label('نوع الطلب')
                    ->live()
                    //->searchable(fn() => self::getEmployee())
                    ->disabled(fn(Forms\Get $get): bool => !(self::getEmployee()))
                    ->options(function (Forms\Get $get, $state) {
                        if (self::getEmployee()) {
                            $allowance_rule = self::$employee->allowance_rule;
                            $allowance_limits = $allowance_rule->limits;
                            return $allowance_limits->pluck('allowance.name', 'id');
                        } else
                        {
                            return ['لا يوجد صلاحيات'];
                        }
                    })
                    ->afterStateUpdated(function (Set $set, Forms\Get $get, $component, $state) {
                        self::$requestId = $state;
                        //self::refreshEmployee($set, $get,$component,$state);
                        self::updateUI($set,$get);
                    })
                    ->columnSpan(4),

                TextInput::make('allowance_amount')
                    ->label('العدد')
                    ->disabled(fn(Forms\Get $get): bool => !(filled($get('requests'))))
                    ->autofocus(fn(Forms\Get $get): bool => (filled($get('requests'))))
                    ->columnSpan(3),

                TextInput::make('allowance_unit')
                    ->label('-')
                    ->readOnly()
                    ->columnSpan(1)

            ])->columns(8)->columnSpanFull(),


            Forms\Components\Group::make()->schema([
                Forms\Components\DatePicker::make('start_date')
                    ->default(now())
                    ->label('تبدأ في تاريخ'),
                Forms\Components\DatePicker::make('end_date')
                    ->default(date_add(now(), date_interval_create_from_date_string('1 days')))
                    ->label('تنتهي في تاريخ'),
                Forms\Components\Textarea::make('notes')
                    ->label('ملحوظات')
                    ->columnSpanFull(),
            ])
                ->columns(2)
                ->columnSpanFull()
                ->disabled(fn(Forms\Get $get): bool => !(filled($get('requests')))),

        ])->columns(2);
    }

    private static function startSessions(): void
    {
        session_start();
        session('employee',null);
        session('request',null);
    }
    private static function endSessions(): void
    {
        session()->forget('employee');
        session()->forget('request');
    }
    public static function form(Form $form): Form
    {
        static::endSessions();
        static::startSessions();

        $allowRule = AllowanceRule::find(1);
        $allowLimits = $allowRule->limits;
        $allowance = $allowLimits[0]->allowance;

        return $form
            ->schema([
                Forms\Components\Hidden::make('code'),
                Forms\Components\Hidden::make('allowance_id'),

                Forms\Components\Group::make()->schema([
                    (new EmployeeAllowanceResource)->createEmployeePhotoSchema(),
                    (new EmployeeAllowanceResource)->createAllowanceLimit(),
                ]),
                Forms\Components\Group::make()->schema([
                    (new EmployeeAllowanceResource)->createSearchSection(),
                    (new EmployeeAllowanceResource)->createRequestForm(),
                ])->columnSpan(2)
            ])->columns(3);
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                //
            ])
            ->filters([
                //
            ])
            ->actions([
                Tables\Actions\EditAction::make(),
            ])
            ->bulkActions([
                Tables\Actions\BulkActionGroup::make([
                    Tables\Actions\DeleteBulkAction::make(),
                ]),
            ]);
    }

    public static function getRelations(): array
    {
        return [
            //
        ];
    }

    public static function getPages(): array
    {
        return [
            'index' => Pages\ListEmployeeAllowances::route('/'),
            'create' => Pages\CreateEmployeeAllowance::route('/create'),
            'edit' => Pages\EditEmployeeAllowance::route('/{record}/edit'),
        ];
    }

    public static function getWidgets(): array
    {
        return [
            FindEmployeeWidget::class
        ];
    }

}

**Comment by maystro** [comment]
I found the mistake
using searchable refreshes the component each request and removes the variable stored in session, that gives null and don't populate the list
